core/src/main/java/edu/wpi/grip/core/sources/CameraSource.java

```diff
@@ -154,8 +154,16 @@ public void start() throws IOException, IllegalStateException {
 
                     frameMat.copyTo(frameOutputSocket.getValue().get());
                     frameOutputSocket.setValue(frameOutputSocket.getValue().get());
-                    long thisMoment = System.currentTimeMillis();
-                    frameRateOutputSocket.setValue(1000 / (thisMoment - lastFrame));
+
+                    final long thisMoment = System.currentTimeMillis();
+                    final long elapsedTime = thisMoment - lastFrame;
+                    if (elapsedTime == 0) {
+                        // This can actually happen: https://github.com/WPIRoboticsProjects/GRIP/issues/199
+                        frameRateOutputSocket.setValue(Double.NaN);
+                    } else {
+                        assert elapsedTime > 0 : "Time has run backwards!";
```

Comment:
You should not use `System.currentTimeMillis()` if going backwards is a problem.  `System.currentTimeMillis()` is non-monotonic, so backwards times are expected.

[`System.nanoTime()` is monotonically increasing](http://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()), so it would be a better choice